improve healthcheck endpoint

zanssa opened the pull request:
Ref: CLOUDDST-9948

zanssa commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,202 @@
+
+    return {
+        "services": services,
+        "software": package_version_check(),
```
For retrieving version of the software, there was a suggestion to represent the version in the requests logs as there would be multiple workers in different VMs. hence, the removal for these functions in the PR.

zanssa commented:
conflicts have been resolved. TravisCI failing because of safety, which be handled in a different story!
I'll squash the commits once this is approved.

lipoja commented on iib/web/status.py:
```diff
@@ -0,0 +1,140 @@
+    """
+    Get status of IIB workers and services that IIB depends on.
+
+    Display the current version of podman, skopeo, opm.
```
This was for previous version of this function, right?

zanssa replied:
Oops! Yeah, I missed it while reviewing.

lipoja commented on iib/web/status.py:
```diff
@@ -0,0 +1,140 @@
+    :param int worker_ping_retries: number of retry/retries on ConnectionError when pinging workers
+    :return: dict with the following keys:
+        "services": list of status info for individual services
+        "software": list of softwares and their versions
```
This was for previous version of this function, right?
Since I see return just "status" and "workers".

lipoja commented on iib/web/status.py:
```diff
@@ -0,0 +1,140 @@
+        raise IIBError("no workers are available")
+
+    return {
+        "status": services,
```
I am not sure if we can change type of "status". Until now it is plain text, however with this change we are returning dictionary. It can break parsing of the response.

lipoja commented on iib/web/api_v1.py:
```diff
@@ -314,23 +314,33 @@ def get_builds():
 
-@api_v1.route('/healthcheck')
-def get_healthcheck():
+@api_v1.route("/status", methods=["GET"])
```
Why are we removing `/healthcheck` endpoint?
This change would cause bumping API version. And I thought we do not want to do it if is not necessary. 
Adding things is OK, but deleting or changing might cause troubles.

zanssa replied:
Oh! yeah so, I changed it to `status` when there were functions for retrieving the version of the pkgs as it would be more suitable than healthchek. Now, that version won't be part of this, I would change it back to the healthcheck.

lipoja commented:
@zanssa I did not check everything since I think it should be modified once again. I think we should not make major changes to current API endpoints. We can add but we should not change or delete if it is not necessary.

midnightercz commented:
Looks good, but lets keep /healthcheck here

zanssa commented:
@lipoja @midnightercz 
so, I am keeping `healthcheck` with the same return type of `"status"`, which is string. If all is good, it will return this:

`{
  "status": "Health check OK"
}`

and I am adding another endpoint  `healthcheck/details` which will contain the new attributes ( services & workers ) 

Thanks for reviewing!

zanssa commented:
in last push: 
- replaced `ok` with `available`

lipoja commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+            log.exception("broker connection failed, exception: %s", OperError)
+            return False, "broker connection failed"
+        except kombu.exceptions.KombuError as err:
+            log.error("an error occurred while trying to connect to the RabbitMQ server: %s", err)
```
Could you please change double quotes " for single quotes '. In all files. Doc-string is exception.
It is noted in README.md -> `* Single quotes are used instead of double quotes`

lipoja replied:
Don't blame me, I like double quotes but we should keep code style in a project.

zanssa replied:
I've just replaced them! :)

lcarva commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+        try:
+            connection.ensure_connection(max_retries=0)
+        except kombu.exceptions.OperationalError as OperError:
+            log.exception("broker connection failed, exception: %s", OperError)
```
I think because you're using `log.exception`, you don't have to include the exception in the log message. The `exception` method will just automatically take the latest exception on the stack. Consider changing this line to: `log.exception("broker connection failed")`

lcarva commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+
+    def add_status(service_name, ok, reason):
+        if short and not ok:
+            raise IIBError(f"{service_name} unavailable: {reason}")
```
If a service is not ok, then the other services are never checked. It may be desirable to always provide the status for every service.

zanssa replied:
in case of short is True, it would only show the status in general (Status: health check failed and the reason) - just like the current healthcheck endpoint so nothing would break as it was suggested & mentioned in this PR. 

As this is a short healthcheck, if one of the services is not working, it should show the status as failed. 

If you wanna see the status of all services, you will need to check `healthcheck/details` - it contains a list of the services ( database & rabbitmq ), workers and their status.

lcarva replied:
Fair point. It's best to stop checking the other services in this case. Thanks for the clarification.

zanssa commented on iib/web/api_v1.py:
```diff
@@ -314,23 +314,33 @@ def get_builds():
+        retval = {"status": "Health check failed", "error": str(error)}
 
-    return flask.jsonify({'status': 'Health check OK'})
+    return flask.jsonify(retval), 200 if retval["status"] == "Health check OK" else 503
```
It does return error. For example, in case db conncection failed 

`"status": "Health check failed", "error": "DATABASE unavailable: Database connection failed"`

may I ask what made you think it won't return any response text?

lcarva replied:
Oh, my bad! I had written the line as:
```python
    return (flask.jsonify(retval), 200) if retval["status"] == "Health check OK" else 503
```

When in fact, it's processed like this (added parenthesis for clarity):
```python
return flask.jsonify(retval), (200 if retval["status"] == "Health check OK" else 503)
```

This doesn't change my comment about the conditional being brittle.

zanssa replied:
I'll modify it! :)

lcarva reviewed (dismissed):
Good job!

pbortlov commented on iib/web/static/api_v1.yaml:
```diff
@@ -210,30 +210,77 @@ paths:
+          description: services and workers are working as expected.
           content:
             application/json:
-              schema:
```
Do we know if nobody uses the schema to automate if IIB works or not? 
I really like to use schema to at least show to have structured output.

zanssa replied:
Can you elaborate more here how this would be used to automate if IIB works or not? 

Isn't ("Health Check OK"  or "Health Check failed") what would be used for validating the status of iib. am I missing something here?

pbortlov replied:
The response  `example: {'status': 'Health check OK'}` is ok but the schema says that it is a json object. Having just an example doesn't say if it is a string or object.

I mean Semaphore used the JSON from the healthcheck to show if IIB is working. I am not sure if there are any similar scripts

zanssa replied:
My understanding is that this file is only for documentation. correct ? 

I know Semaphore would be shutdown soon, and I am not familiar how it checks if iib is working or not based on the schema. So, I am looping @midnightercz as he knows more and can tell if it would break something or not. wdyt Jindrich?

lipoja replied:
@zanssa If we accept this change this new/modified endpoint will be the only one that does not have it documented.
And you are saying "only for documentation"? I mean - you never relied on documentation? Are you always looking to the code to see how to work with any API that you are using?

I agree with @pbortlov that it is good to have it documented. Why we should lower our standards with this PR?

Could you add the schema here, please?

zanssa replied:
Oh! I think the change is already approved :)

There is a difference between using the documentation as instructions and using the schema there to test if iib is working on not.

What is your standard that is lowered with this PR?

lipoja replied:
For me the standard for every documented endpoint is to have this schema - therefore I would assume this one should have it as well.

yashvardhannanavati replied:
@zanssa It shouldn't matter if the change is approved. One person might miss something which the other would notice. Please do address comments from all reviewers. We shouldn't merge a change unless concerns of all our guild members are resolved. 

fwiw, I agree we shouldn't cut corners in documentation and keep it as accurate as possible. One use case could be if another service uses IIB's endpoint to check its status and assumes it's going to be string instead of JSON from the documentation.

zanssa replied:
@yashvardhannanavati I addressed the comments from all and I even took into consideration what I was asked in gchat instead of here, and despite that some of them are not really necessary or even enforced in the repo.

I would expect some change requests and I don't really mind making the changes but it would be nice if those change requests are mentioned after reviewing once or twice; not one change every few days. This is just slowing the progress and not helping! :)

pbortlov commented on tests/test_web/test_api_v1.py:
```diff
@@ -2103,3 +2089,27 @@ def test_get_build_related_bundles_not_configured(client, db, minimal_request_re
+
+@pytest.mark.parametrize('error', [None, 'something is wrong'])
+@mock.patch('iib.web.api_v1.status')
+def test_get_status_short(mock_status, error, client):
```
Could we have tests for success and failure separated? The good practice is to use separate functions if you want to test `if` statement
- test_get_status_short_success
- test_get_status_short_failure

zanssa replied:
aha. I'll keep that in mind when creating tests in the future!

lipoja replied:
On the other hand, if you do it in this PR we do not have to modify it in a future, right?
You can have a look to this article: https://www.nerdwallet.com/blog/engineering/5-pytest-best-practices/ 
From the article:
> Parametrize the same behavior, have different tests for different behaviors
> 
> tl;dr: Parametrize when asserting the same behavior with various inputs and expected outputs. Make separate tests for distinct behaviors.

Correct me if I am wrong, but I think it is different behavior when test ends with error and when it is successful.

zanssa replied:
Yeah, if it is that important, we can do it in the future! I think the tests in test_healthcheck file is enough. Plus, I don't have the capacity for further requests. :)

yashvardhannanavati replied:
@zanssa Could you please make the change on this suggestion. I agree success and failure should be separate tests.

pbortlov commented on tests/test_web/test_healthcheck.py:
```diff
@@ -0,0 +1,240 @@
+    mock_rabbitmq_connection_check.return_value = (True, None)
+    mock_workers_status.return_value = workers_status
+
+    if short and not expected_result:
```
Could add to parametrize a variable error and error_message instead of adding `if` statement?

yashvardhannanavati replied:
@zanssa This comment was resolved without a response. Could you please explain why you prefer to keep it this way?

Ideally, it would be nice if you could split this into two different tests instead of one. But I am willing to look the other way

lipoja commented on tests/test_web/test_healthcheck.py:
```diff
@@ -0,0 +1,240 @@
+            healthcheck.status(short=True)
+        return
+
+    result = healthcheck.status(short=False)
```
should not this `(short=False)` be `(short=short)` instead? Or why do you have the short parametrized?

zanssa replied:
Can you explain to me why do you think it should be `short=short` and not `short=False` ?

lipoja replied:
Please, do not resolve conversation opened by others. I mean why not to use the variable from parametrize?

zanssa replied:
I've always done it this way, resolving after addressing the reviewer's concerns. I marked it as resolved as there is no change required here. If you see there is another approach, that's great, but I don't think I'll change these tests :)

yashvardhannanavati replied:
@zanssa I agree, we should use the variable value instead of hard coded values here since you have parameterized the value already. 

It's very easy for someone else altering your test to miss the hard coded value. +1 on using the variable name.

lipoja commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,202 @@
+    else:
+        workers = []
+
+    any_worker_ok = any(worker["ok"] for worker in workers)
```
@yashvardhannanavati could you help me remember the conclusion on this topic. Thanks!

yashvardhannanavati replied:
Thinking about it now, we should probably fail even if one of the workers is no working. Here's why, CVP might have 6 workers and 1 of them might be down. But CVP doesn't know which of its request will be routed to that 1 unstable worker giving them a bad experience. We as a team should be proactive to react to it rather than ignoring it as a warning. 

So I think simply listing the status of the workers as @zanssa mentioned is sufficient, let's just make the health check fail on that. @zanssa I apologize for this last minute change but it just occured to me. I would be ok, if you want to cover this change in a separate ticket and keep this PR working as we discussed previously

lipoja replied:
@yashvardhannanavati Thanks for this idea, it is easier to implement then the solution I had in mind.

zanssa commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,202 @@
+    :param int retries: number of retry/retries on ConnectionError when pinging workers
+    :return: list of workers status information (empty if no replies)
+    """
+    replies = _ping_celery_workers(retries=retries)
```
I think I addressed this before here: https://github.com/release-engineering/iib/pull/316#discussion_r737681926 

you could also refer to the story description, and the last comment! I just updated the story to point out that version of software is removed from this and healthcheck/short is replaced by healthcheck/details!

yashvardhannanavati commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+    """
+    Check if the database connection is working.
+
+    :return: tuple (ok: bool, reason: str or None)
```
could you add `:raises` and `rtype` to the doc string.

yashvardhannanavati replied:
Please refer other existing doc strings and keep the format uniform

zanssa replied:
Sure! I can do that. :) 

There was a request to add these in another file, which I did. But I don't see this implemented in all files. That's why I didn't add it here; same thing with single & double quotes. I do understand that this is not a reason to not follow the format, but, if we want to enforce a rule, we should always apply it to all files, not just decide to enforce it when it suits us.

lipoja replied:
I agree with you Zainab! It should be in all files. However I do not know how to automate these checks. So it might happen that developer forget to add it and it slips through code-review. 

If you see missing `:raises` and `rtype` feel free to create separate PR for that. Same with single/double quotes.

Or if you know a way how to automate this checks, please share it with us. It would be great to have these checks in our test suit - so it is really enforced.

zanssa replied:
@lipoja you agreed and asked for it which means you believe it is important. but I don't see this. :) 

I reminded you about the code standard you have a concern about here and at the same time forgot about in one of your PRs. You changed it there and then sent another PR with the same missing code standard. 

I am not against it but let's all comply with the same standards. 

There is flake8 extension to lint for quotes (flake8-quotes) 

You could install it by pip and test it in your local. 

`pip install flake8-quotes
`

as an example, run this cmd: `flake8 --inline-quotes "'" iib/workers/tasks/utils.py`

the result should be similar to this: 

`iib/workers/tasks/utils.py:73:26: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:280:15: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:280:32: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:280:54: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:333:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:334:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:335:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:336:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:337:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:356:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:357:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:358:9: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:379:38: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:379:52: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:982:15: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:982:22: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:982:31: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1007:38: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:11: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:25: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:35: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:56: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:66: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1018:82: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1032:39: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1066:43: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1067:35: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1094:25: Q000 Double quotes found but single quotes preferred
iib/workers/tasks/utils.py:1094:54: Q000 Double quotes found but single quotes preferred`

there is also support for multiline. For ref: https://github.com/zheller/flake8-quotes

yashvardhannanavati replied:
@zanssa Thanks for pointing this out to us. I agree, it should be in all files. Maybe we missed it. Let's file an issue to add this enforcing? In the meanwhile, let's stay consistent going forward. I would really appreciate if you could make the change. Sorry to be nitpicking

lipoja replied:
@zanssa Thanks for finding this out! Yes, as I said in https://github.com/release-engineering/iib/pull/316#discussion_r748579261 I do not like it and I prefer double quotes. Therefore I have hard time to write code with single quotes as well. It would be great if we can enforce it.

yashvardhannanavati commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+    """
+    Check if RabbitMQ connection is working.
+
+    :param str broker_url: the AMQP(S) URL to connect to RabbitMQ
```
Same as above. The doc string needs to be fixed.

yashvardhannanavati commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+    """
+    Attempt to ping worker, retry on ConnectionError.
+
+    :param int retries: number of retry/retries before returning an empty response
```
Same as above.

yashvardhannanavati commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+    """
+    Ping workers, check received replies.
+
+    :param int retries: number of retry/retries on ConnectionError when pinging workers
```
Same as above.

yashvardhannanavati commented on iib/web/healthcheck.py:
```diff
@@ -0,0 +1,138 @@
+    :return: dict with the following keys:
+        "services": list of status info for individual services
+        "workers": list of status info for individual workers
+    :raises IIBError: if short is True and a problem is found
```
Missing `:rtype`

zanssa commented:
> @zanssa It shouldn't matter if the change is approved. One person might miss something which the other would notice. Please do address comments from all reviewers. We shouldn't merge a change unless concerns of all our guild members are resolved.

> fwiw, I agree we shouldn't cut corners in documentation and keep it as accurate as possible. One use case could be if another service uses IIB's endpoint to check its status and assumes it's going to be string instead of JSON from the documentation.


@yashvardhannanavati I addressed the comments from all and I even took into consideration what I was asked in gchat instead of here, and despite that some of them are not really necessary or even enforced in the repo.

I would expect some change requests and I don't really mind making the changes but it would be nice if those change requests are mentioned after reviewing once or twice; not one change every few days. This is just slowing the progress and not helping! :)

lipoja commented on iib/web/static/api_v1.yaml:
```diff
@@ -275,11 +275,15 @@ paths:
           content:
             application/json:
+              schema:
+                type: object
```
Thank you for adding the schema, I really appreciate the effort!
However the schema was nested and there were also properties set for given object which is still missing.
Could you add correct properties for both responses (200 and 500)?

Thank you!

yashvardhannanavati commented:
> > @zanssa It shouldn't matter if the change is approved. One person might miss something which the other would notice. Please do address comments from all reviewers. We shouldn't merge a change unless concerns of all our guild members are resolved.
> 
> > fwiw, I agree we shouldn't cut corners in documentation and keep it as accurate as possible. One use case could be if another service uses IIB's endpoint to check its status and assumes it's going to be string instead of JSON from the documentation.
> 
> @yashvardhannanavati I addressed the comments from all and I even took into consideration what I was asked in gchat instead of here, and despite that some of them are not really necessary or even enforced in the repo.
> 
> I would expect some change requests and I don't really mind making the changes but it would be nice if those change requests are mentioned after reviewing once or twice; not one change every few days. This is just slowing the progress and not helping! :)

I agree we need to be quick in reviewing and giving feedback. But at the same time, it's also true that whether the comment comes early or late, if it's enforcing good coding practices we should act on them. I understand your concern and we should definitely address it in the guild meeting next week.

pbortlov commented:
@release-engineering/exd-guild-hello-operator Could we do a final review?
